src/Microsoft.TestPlatform.CrossPlatEngine/AttachmentsProcessing/DataCollectorAttachmentsProcessorsFactory.cs

```diff
@@ -0,0 +1,90 @@
+        private const string CoverageFriendlyName = "Code Coverage";
+
+        public IReadOnlyDictionary<string, IConfigurableDataCollectorAttachmentProcessor> Create(InvokedDataCollector[] invokedDataCollectors)
+        {
+            IDictionary<string, Tuple<string, IConfigurableDataCollectorAttachmentProcessor>> datacollectorsAttachmentsProcessors = new Dictionary<string, Tuple<string, IConfigurableDataCollectorAttachmentProcessor>>();
+            bool addCodeCoverageAttachmentProcessors = true;
+
+            if (invokedDataCollectors != null)
+            {
+                // We order files by filename descending so in case of the same collector from the same nuget but with different versions, we'll run the newer version.
+                // i.e. C:\Users\xxx\.nuget\packages\coverlet.collector
+                // /3.0.2
+                // /3.0.3
+                // /3.1.0
+                foreach (var invokedDataCollector in invokedDataCollectors.OrderByDescending(d => d.FilePath))
+                {
+                    EqtTrace.Info($"DataCollectorAttachmentsProcessorsFactory: Analyzing data collector attachment processor Uri: {invokedDataCollector.Uri} AssemblyQualifiedName: {invokedDataCollector.AssemblyQualifiedName} FilePath: {invokedDataCollector.FilePath} HasAttachmentProcessor: {invokedDataCollector.HasAttachmentProcessor}");
+
+                    // We'll merge using only one AQN in case of more "same processors" in different assembly.
+                    if (!invokedDataCollector.HasAttachmentProcessor || datacollectorsAttachmentsProcessors.ContainsKey(invokedDataCollector.AssemblyQualifiedName))
+                    {
+                        continue;
+                    }
+
+                    var dataCollectorExtensionManager = DataCollectorExtensionManager.Create(invokedDataCollector.FilePath, TestSessionMessageLogger.Instance);
+                    var dataCollectorExtension = dataCollectorExtensionManager.TryGetTestExtension(invokedDataCollector.Uri);
+                    if (dataCollectorExtension?.Metadata.HasAttachmentProcessor == true)
+                    {
+                        Type attachmentProcessorType = ((DataCollectorConfig)dataCollectorExtension.TestPluginInfo).AttachmentsProcessorType;
+                        IConfigurableDataCollectorAttachmentProcessor dataCollectorAttachmentProcessorInstance = null;
+                        try
+                        {
+                            dataCollectorAttachmentProcessorInstance = TestPluginManager.CreateTestExtension<IConfigurableDataCollectorAttachmentProcessor>(attachmentProcessorType);
+                        }
+                        catch (Exception ex)
+                        {
+                            EqtTrace.Error($"DataCollectorAttachmentsProcessorsFactory: Failed during the creation of data collector attachment processor '{attachmentProcessorType.AssemblyQualifiedName}'\n{ex}");
```

Comment:
>EqtTrace.Error [](http://example.com/codeflow?start=28&length=14)

curious, where do these get printed? Do they go into the test platform diagnostic log file? How about to the output->Tests pane if it is driven through VS...

Reply:
They will go to diagnostic files.

> How about to the output->Tests pane if it is driven through VS...

At the moment looks like this messages are not sent to tests pane in VS. I'll verify the output story there.

Reply:
I've added message logger, it will flow to VS thanks to ITestRunAttachmentsProcessingEventsHandler